modules/javafx.base/src/main/java/com/sun/javafx/binding/ListenerList.java

```diff
@@ -0,0 +1,195 @@
+
+            // skip if this listener was removed during a notification:
+            if (listener == null) {
+                continue;
+            }
+
+            // communicate to a lower level loop (if triggered) how many listeners were notified so far:
+            progress = i;
+
+            // call invalidation listener (and perhaps a nested notification):
+            callInvalidationListener(observableValue, listener);
+        }
+
+        int changeListenersSize = changeListenersSize();
+        int maxChanges = wasLocked ? Math.min(initialProgress + 1 - invalidationListenersSize, changeListenersSize) : changeListenersSize;
+
+        T newValue = null;
+
+        progress = NESTED_NOTIFICATION_COMPLETED;  // reset progress to ensure latest value is queried at least once
+
+        for (int i = 0; i < maxChanges; i++) {
+            ChangeListener<T> listener = getChangeListener(i);
+
+            // skip if this listener was removed during a notification:
+            if (listener == null) {
+                continue;
+            }
+
+            // only get the latest value if this is the first loop or a nested notification occurred
+            if (progress < 0) {
+                newValue = observableValue.getValue();
+
+                valueObtained(newValue);
+
+                if (Objects.equals(newValue, oldValue)) {
+                    progress = NESTED_NOTIFICATION_ABORTED;  // Indicate an early exit before notifying all listeners intended at this level
+
+                    return wasLocked ? false : unlock();
+                }
+            }
+
+            // communicate to a lower level loop (if triggered) how many listeners were notified so far:
+            progress = i + invalidationListenersSize;
+
+            // call change listener (and perhaps a nested notification updating progress field):
+            callChangeListener(observableValue, listener, oldValue, newValue);
+
+            if (progress == NESTED_NOTIFICATION_ABORTED) {
+
+                /*
+                 * Non-convergence detected: The listener just notified above of value X
+                 * triggered a change to Y. The nested notification loop informing earlier
+                 * listeners of Y was aborted because another listener changed the value
+                 * back to X.
+                 *
+                 * Since listeners are only called when the old value is the last provided new
+                 * value, and not when old == new, the listener that forced Y may incorrectly
+                 * assume the value is still Y, leading to potential inconsistencies.
+                 * Repeated changes between X and Y would normally cause a StackOverflowError.
+                 * This conflicting listener behavior will be reported to the user:
+                 */
+
+                throw new StackOverflowError("non-converging value detected in value modifying listeners on " + observableValue + "; original value was: " + oldValue);
```

Comment:
We shouldn't throw errors, even if that's what the old implementation would implicitly do (not that it matters, no sane application would ever try to catch a `StackOverflowError` as part of its normal control flow). I think a `RuntimeException` is in order.

Also maybe make the message a bit clearer:
`<observableValue> was changed during the invocation of multiple listeners, but the new values diverge [original value: <oldValue>]`

Reply:
Is it possible to ignore the listener that incompatibly changed the value from Y back to X?